sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/BufferSendState.scala

```diff
@@ -0,0 +1,232 @@
+  private[this] var blockRanges: Seq[BlockRange[SendBlock]] = windowedBlockIterator.next()
+
+  private[this] var acquiredBuffs: Seq[RangeBuffer] = Seq.empty
+
+  def getTransferRequest: TransferRequest = synchronized {
+    transferRequest
+  }
+
+  def hasNext: Boolean = synchronized { !markedDone }
+
+  private[this] def freeBounceBuffers(): Unit = {
+    sendBounceBuffers.close()
+  }
+
+  def getTransferResponse(): RefCountedDirectByteBuffer = synchronized {
+    new RefCountedDirectByteBuffer(
+      ShuffleMetadata.buildBufferTransferResponse(bufferMetas))
+  }
+
+  override def close(): Unit = synchronized {
+    require(markedDone)
+    if (isClosed){
+      throw new IllegalStateException("ALREADY CLOSED!")
+    }
+    isClosed = true
+    freeBounceBuffers()
+    request.close()
+  }
+
+  case class RangeBuffer(
+      range: BlockRange[SendBlock], rapidsBuffer: RapidsBuffer)
+      extends AutoCloseable {
+    override def close(): Unit = {
+      rapidsBuffer.close()
+    }
+  }
+
+  def next(): AddressLengthTag = synchronized {
+    require(acquiredBuffs.isEmpty,
+      "Called next without calling `releaseAcquiredToCatalog` first")
+
+    var bounceBuffToUse: MemoryBuffer = null
+    var buffOffset = 0L
+
+    val buffsToSend = try {
+      if (!markedDone) {
+        var deviceBuffs = 0L
+        var hostBuffs = 0L
+        acquiredBuffs = blockRanges.safeMap { blockRange =>
+          val transferRequest =
+            blockRange.block.bufferTransferRequest
+
+          // we acquire these buffers now, and keep them until the caller releases them
+          // using `releaseAcquiredToCatalog`
+          closeOnExcept(
+            requestHandler.acquireShuffleBuffer(transferRequest.bufferId())) { rapidsBuffer =>
+            //these are closed later, after we synchronize streams
+            rapidsBuffer.storageTier match {
+              case StorageTier.DEVICE =>
+                deviceBuffs += blockRange.rangeSize()
+              case _ => // host/disk
+                hostBuffs += blockRange.rangeSize()
+            }
+            RangeBuffer(blockRange, rapidsBuffer)
+          }
+        }
+
+        logDebug(s"Occupancy for bounce buffer is [device=${deviceBuffs}, host=${hostBuffs}] Bytes")
+
+        bounceBuffToUse = if (deviceBuffs >= hostBuffs || hostBounceBuffer == null) {
+          deviceBounceBuffer.buffer
+        } else {
+          hostBounceBuffer.buffer
+        }
+
+        acquiredBuffs.foreach { case RangeBuffer(blockRange, rapidsBuffer) =>
+          require(blockRange.rangeSize() <= bounceBuffToUse.getLength - buffOffset)
+          withResource(rapidsBuffer.getMemoryBuffer) { memBuff =>
+            bounceBuffToUse match {
+              case _: HostMemoryBuffer =>
+                //TODO: HostMemoryBuffer needs the same functionality that
+                // DeviceMemoryBuffer has to copy from/to device/host buffers
+                CudaUtil.copy(
```

Comment:
This copy is not using `serverStream`?

Reply:
We don't have a `HostMemoryBuffer` function that will take the stream and offset for this copy, so for now I left these as synchronous. We'll have to change CUDF and will update this code.

Reply:
Added it here: https://github.com/rapidsai/cudf/issues/6697